File: app/src/main/java/com/example/android/simpledice/ui/DKeyboard.java
package com.example.android.simpledice.ui;

import android.content.Context;
import android.os.Handler;
import android.support.constraint.ConstraintLayout;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.inputmethod.InputConnection;
import android.widget.Button;
import android.widget.ImageButton;

import com.example.android.simpledice.R;

public class DKeyboard extends ConstraintLayout implements View.OnClickListener, View.OnLongClickListener{

    private Button mOneButton, mTwoButton, mThreeButton, mFourButton,
            mFiveButton, mSixButton, mSevenButton, mEightButton,
            mNineButton, mZeroButton, mDButton, mPlusButton,
            mMinusButton, mEnterButton;
    private ImageButton mDeleteButton;
    private final int MILLIS_BETWEEN_CHARACTER_CHANGE = 50;

    private SparseArray<String> keyValues = new SparseArray<>();
    private InputConnection mInputConnection;
    private Animation.AnimationListener mExitAnimationListener;
    private Animation.AnimationListener mEnterAnimationListener;

    public DKeyboard(Context context) {
        super(context);
        init(context);
    }

    public DKeyboard(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public DKeyboard(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        LayoutInflater.from(context).inflate(R.layout.d_keyboard, this, true);

        //Assign view
        mOneButton = (Button) findViewById(R.id.one_button);
        mTwoButton = (Button) findViewById(R.id.two_button);
        mThreeButton = (Button) findViewById(R.id.three_button);
        mFourButton = (Button) findViewById(R.id.four_button);
        mFiveButton = (Button) findViewById(R.id.five_button);
        mSixButton = (Button) findViewById(R.id.six_button);
        mSevenButton = (Button) findViewById(R.id.seven_button);
        mEightButton = (Button) findViewById(R.id.eight_button);
        mNineButton = (Button) findViewById(R.id.nine_button);
        mZeroButton = (Button) findViewById(R.id.zero_button);
        mPlusButton = (Button) findViewById(R.id.plus_button);
        mMinusButton = (Button) findViewById(R.id.minus_button);
        mDButton = (Button) findViewById(R.id.d_button);
        mDeleteButton = (ImageButton) findViewById(R.id.delete_button);
        mEnterButton = (Button) findViewById(R.id.enter_button);

        //Assign clickListeners
        mOneButton.setOnClickListener(this);
        mTwoButton.setOnClickListener(this);
        mThreeButton.setOnClickListener(this);
        mFourButton.setOnClickListener(this);
        mFiveButton.setOnClickListener(this);
        mSixButton.setOnClickListener(this);
        mSevenButton.setOnClickListener(this);
        mEightButton.setOnClickListener(this);
        mNineButton.setOnClickListener(this);
        mZeroButton.setOnClickListener(this);
        mPlusButton.setOnClickListener(this);
        mMinusButton.setOnClickListener(this);
        mDButton.setOnClickListener(this);
        mDeleteButton.setOnClickListener(this);
        mEnterButton.setOnClickListener(this);

        //Assign longClickListeners to everything but Enter
        mOneButton.setOnLongClickListener(this);
        mTwoButton.setOnLongClickListener(this);
        mThreeButton.setOnLongClickListener(this);
        mFourButton.setOnLongClickListener(this);
        mFiveButton.setOnLongClickListener(this);
        mSixButton.setOnLongClickListener(this);
        mSevenButton.setOnLongClickListener(this);
        mEightButton.setOnLongClickListener(this);
        mNineButton.setOnLongClickListener(this);
        mZeroButton.setOnLongClickListener(this);
        mPlusButton.setOnLongClickListener(this);
        mMinusButton.setOnLongClickListener(this);
        mDButton.setOnLongClickListener(this);
        mDeleteButton.setOnLongClickListener(this);

        //Populate SparseArray with key value pairs
        keyValues.put(R.id.one_button, "1");
        keyValues.put(R.id.two_button, "2");
        keyValues.put(R.id.three_button, "3");
        keyValues.put(R.id.four_button, "4");
        keyValues.put(R.id.five_button, "5");
        keyValues.put(R.id.six_button, "6");
        keyValues.put(R.id.seven_button, "7");
        keyValues.put(R.id.eight_button, "8");
        keyValues.put(R.id.nine_button, "9");
        keyValues.put(R.id.zero_button, "0");
        keyValues.put(R.id.plus_button, "+");
        keyValues.put(R.id.minus_button, "-");
        keyValues.put(R.id.d_button, "d");

        //Assign the animation listeners
        mExitAnimationListener = new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {
                //Do nothing
            }

            @Override
            public void onAnimationEnd(Animation animation) {
                DKeyboard.this.clearAnimation(); //Clear the animation to allow for subsequent adjusting of view's visibility
                DKeyboard.this.setVisibility(View.GONE); //Necessary, lest the view be clickable
            }

            @Override
            public void onAnimationRepeat(Animation animation) {
                //Do nothing
            }
        };

        mEnterAnimationListener = new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {
                //Do nothing
            }

            @Override
            public void onAnimationEnd(Animation animation) {
                DKeyboard.this.clearAnimation(); //Clear the animation to allow for subsequent adjusting of the view's visibility
            }

            @Override
            public void onAnimationRepeat(Animation animation) {
                //Do nothing
            }
        };
    }

    @Override
    public void onClick(View v) {
        if (mInputConnection == null) return;

        int viewId = v.getId();
        switch (viewId) {
            case (R.id.delete_button): { //Special delete logic

                deleteText();

                break;
            }
            case(R.id.enter_button) : { //Special enter logic
                DKeyboard.this.executeExitAnimation();
                break;
            }
            default : { //Other buttons simply append the value of the button, accessed from keyValues
                String value = keyValues.get(viewId);
                mInputConnection.commitText(value, 1);
            }
        }
    }

    @Override
    public boolean onLongClick(final View v) {
        if (mInputConnection == null) return true;
        final int viewId = v.getId();
        switch(viewId) {
            case(R.id.delete_button) : { //Special delete logic, continuous delete
                final Handler handler = new Handler();
                final Runnable runnable = new Runnable() {
                    @Override
                    public void run() {
                        if (v.isPressed()){
                            deleteText();
                            handler.postDelayed(this, MILLIS_BETWEEN_CHARACTER_CHANGE);
                        }
                    }
                };
                handler.postDelayed(runnable, MILLIS_BETWEEN_CHARACTER_CHANGE);
                return true;
            }
            default: { //Otherwise, continuous insertion of character
                final Handler handler = new Handler();
                final Runnable runnable = new Runnable() {
                    @Override
                    public void run() {
                        if (v.isPressed()){
                            String value = keyValues.get(viewId);
                            mInputConnection.commitText(value, 1);
                            handler.postDelayed(this, MILLIS_BETWEEN_CHARACTER_CHANGE);
                        }
                    }
                };
                handler.postDelayed(runnable, MILLIS_BETWEEN_CHARACTER_CHANGE);
                return true;
            }
        }
    }

    private void deleteText(){
        CharSequence selectedText = mInputConnection.getSelectedText(0);

        if (TextUtils.isEmpty(selectedText)) { //If not highlighted, delete 1 before, 0 after
            mInputConnection.deleteSurroundingText(1, 0);
        } else { //Simply replace highlighted text with empty string
            mInputConnection.commitText("", 1);
        }

    }

    public void setInputConnection(InputConnection inputConnection) {
        mInputConnection = inputConnection;
    }

    public void executeExitAnimation(){
        Animation exitAnimation = AnimationUtils.loadAnimation(getContext(), R.anim.bottom_exit);
        exitAnimation.setAnimationListener(mExitAnimationListener);
        DKeyboard.this.startAnimation(exitAnimation);
    }

    public void executeEnterAnimation(){
        DKeyboard.this.setVisibility(View.VISIBLE);
        Animation enterAnimation = AnimationUtils.loadAnimation(getContext(), R.anim.bottom_enter);
        enterAnimation.setAnimationListener(mEnterAnimationListener);
        DKeyboard.this.startAnimation(enterAnimation);
    }
}

File: SimpleDice/app/src/main/java/com/gmail/maxfixsoftware/simpledice/ui/DKeyboard.kt
package com.gmail.maxfixsoftware.simpledice.ui

import android.content.Context
import android.os.Handler
import android.text.TextUtils
import android.util.AttributeSet
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.view.inputmethod.InputConnection
import androidx.constraintlayout.widget.ConstraintLayout
import com.gmail.maxfixsoftware.simpledice.R
import com.gmail.maxfixsoftware.simpledice.utils.Constants.MILLIS_BETWEEN_CHARACTER_CHANGE
import kotlinx.android.synthetic.main.d_keyboard.view.*

class DKeyboard : ConstraintLayout, View.OnClickListener, View.OnLongClickListener {

    private val keyValues = SparseArray<String>()
    private var mInputConnection: InputConnection? = null
    private var mExitAnimationListener: Animation.AnimationListener? = null
    private var mEnterAnimationListener: Animation.AnimationListener? = null

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    private fun init(context: Context) {
        LayoutInflater.from(context).inflate(R.layout.d_keyboard, this, true)

        //Assign clickListeners
        one_button.setOnClickListener(this)
        two_button.setOnClickListener(this)
        three_button.setOnClickListener(this)
        four_button.setOnClickListener(this)
        five_button.setOnClickListener(this)
        six_button.setOnClickListener(this)
        seven_button.setOnClickListener(this)
        eight_button.setOnClickListener(this)
        nine_button.setOnClickListener(this)
        zero_button.setOnClickListener(this)
        plus_button.setOnClickListener(this)
        minus_button.setOnClickListener(this)
        d_button.setOnClickListener(this)
        delete_button.setOnClickListener(this)
        enter_button.setOnClickListener(this)

        //Assign longClickListeners to everything but Enter
        one_button.setOnLongClickListener(this)
        two_button.setOnLongClickListener(this)
        three_button.setOnLongClickListener(this)
        four_button.setOnLongClickListener(this)
        five_button.setOnLongClickListener(this)
        six_button.setOnLongClickListener(this)
        seven_button.setOnLongClickListener(this)
        eight_button.setOnLongClickListener(this)
        nine_button.setOnLongClickListener(this)
        zero_button.setOnLongClickListener(this)
        plus_button.setOnLongClickListener(this)
        minus_button.setOnLongClickListener(this)
        d_button.setOnLongClickListener(this)
        delete_button.setOnLongClickListener(this)

        //Populate SparseArray with key value pairs
        keyValues.put(R.id.one_button, "1")
        keyValues.put(R.id.two_button, "2")
        keyValues.put(R.id.three_button, "3")
        keyValues.put(R.id.four_button, "4")
        keyValues.put(R.id.five_button, "5")
        keyValues.put(R.id.six_button, "6")
        keyValues.put(R.id.seven_button, "7")
        keyValues.put(R.id.eight_button, "8")
        keyValues.put(R.id.nine_button, "9")
        keyValues.put(R.id.zero_button, "0")
        keyValues.put(R.id.plus_button, "+")
        keyValues.put(R.id.minus_button, "-")
        keyValues.put(R.id.d_button, "d")

        //Assign the animation listeners
        mExitAnimationListener = object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {
                //Do nothing
            }

            override fun onAnimationEnd(animation: Animation) {
                this@DKeyboard.clearAnimation() //Clear the animation to allow for subsequent adjusting of view's visibility
                this@DKeyboard.visibility = View.GONE //Necessary, lest the view be clickable
            }

            override fun onAnimationRepeat(animation: Animation) {
                //Do nothing
            }
        }

        mEnterAnimationListener = object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {
                //Do nothing
            }

            override fun onAnimationEnd(animation: Animation) {
                this@DKeyboard.clearAnimation() //Clear the animation to allow for subsequent adjusting of the view's visibility
            }

            override fun onAnimationRepeat(animation: Animation) {
                //Do nothing
            }
        }
    }

    override fun onClick(v: View) {
        if (mInputConnection == null) return

        val viewId = v.id
        when (viewId) {
            R.id.delete_button -> { //Special delete logic

                deleteText()
            }
            R.id.enter_button -> { //Special enter logic
                this@DKeyboard.executeExitAnimation()
            }
            else -> { //Other buttons simply append the value of the button, accessed from keyValues
                val value = keyValues.get(viewId)
                mInputConnection!!.commitText(value, 1)
            }
        }
    }

    override fun onLongClick(v: View): Boolean {
        if (mInputConnection == null) return true
        val viewId = v.id
        when (viewId) {
            R.id.delete_button -> { //Special delete logic, continuous delete
                val handler = Handler()
                val runnable = object : Runnable {
                    override fun run() {
                        if (v.isPressed) {
                            deleteText()
                            handler.postDelayed(this, MILLIS_BETWEEN_CHARACTER_CHANGE.toLong())
                        }
                    }
                }
                handler.postDelayed(runnable, MILLIS_BETWEEN_CHARACTER_CHANGE.toLong())
                return true
            }
            else -> { //Otherwise, continuous insertion of character
                val handler = Handler()
                val runnable = object : Runnable {
                    override fun run() {
                        if (v.isPressed) {
                            val value = keyValues.get(viewId)
                            mInputConnection!!.commitText(value, 1)
                            handler.postDelayed(this, MILLIS_BETWEEN_CHARACTER_CHANGE.toLong())
                        }
                    }
                }
                handler.postDelayed(runnable, MILLIS_BETWEEN_CHARACTER_CHANGE.toLong())
                return true
            }
        }
    }

    private fun deleteText() {
        val selectedText = mInputConnection!!.getSelectedText(0)

        if (TextUtils.isEmpty(selectedText)) { //If not highlighted, delete 1 before, 0 after
            mInputConnection!!.deleteSurroundingText(1, 0)
        } else { //Simply replace highlighted text with empty string
            mInputConnection!!.commitText("", 1)
        }

    }

    fun setInputConnection(inputConnection: InputConnection) {
        mInputConnection = inputConnection
    }

    fun executeExitAnimation() {
        val exitAnimation = AnimationUtils.loadAnimation(context, R.anim.bottom_exit)
        exitAnimation.setAnimationListener(mExitAnimationListener)
        this@DKeyboard.startAnimation(exitAnimation)
    }

    fun executeEnterAnimation() {
        this@DKeyboard.visibility = View.VISIBLE
        val enterAnimation = AnimationUtils.loadAnimation(context, R.anim.bottom_enter)
        enterAnimation.setAnimationListener(mEnterAnimationListener)
        this@DKeyboard.startAnimation(enterAnimation)
    }
}
